### ISSUE TITLE
Job update fails with error (Exec[reload-jenkins] not in catalog)

### ISSUE THREAD

**Issue by ofalk** [jenkinsci/puppet-jenkins#294]
Hi!

I pulled the latest source today and now I get the following error:

"""
Could not retrieve catalog from remote server: Error 400 on SERVER: Invalid relationship: Exec[jenkins update-job puppet] { notify => Exec[reload-jenkins] }, because Exec[reload-jenkins] doesn't seem to be in the catalog
"""

A bug or a misconfiguration on my side?

**Comment by jhoblitt** [comment]
This is likely a regression caused by.  https://github.com/jenkinsci/puppet-jenkins/commit/979cb012c12bc80306a88465f8cd78f43b32f70d  Do you have `class { 'jenkins': cli => false` }`?

**Comment by ofalk** [comment]
I have no ::cli stuff defined at all :-/
Shall I try to define it? Earliest tomorrow...

**Comment by ofalk** [comment]
Adding cli => true to my jenkins class resolved the error!
So maybe you want to add this as a default and not allow disabling it!?

**Comment by jhoblitt** [comment]
I suspect that `jenkins::cli` needs to include `jenkins::cli::reload` to fix the regression.  I'll need to do some testing to make sure that doesn't reintroduce the dependency loop for `jenkins::cli::exec` that commit resolved.

**Comment by nand0p** [comment]
this change broke my CD. fix: sed -i 's/include\ jenkins::cli/include\ jenkins::cli\ninclude\ jenkins::cli::reload/' /etc/puppet/modules/jenkins/manifests/job/present.pp

**Comment by jhoblitt** [comment]
@ofalk / @nand0p It turns out there are multiple issues here.

1.  This module implements "the anchor" pattern for resource ordering which is not  enabled for usage of the "cli" unless `class { 'jenkins': cli => true }`.  As the module is currently designed, it should really disallow the usage of any class/define that uses the cli.  We need to try to support the defacto API but a breaking change will likely be needed to fix this properly.
2.  Even if `cli => true`, the defined types are not safely ordered not are they contained but the `Anchor[jenkins::end]` resource.
3. Due to \1, none of the unit tests inspect resource ordering
4. There are no acceptance tests that cover any of the defined types.

I have attempted to fix the inclusion problem reported in this issue, fix \2 which I just described, and add a simple acceptance test for jenkins::job.  That later appears to be non-idempotent with jenkins 1.1613.

It would be fantastic if either of you would be willing to contribute acceptance tests for the other interface classes and types.

@rtyler #298 should fix this issue, so there's likely some urgency to get a patch release out, but I'd advise careful review for fear of introducing new ordering problems.
